src/readers/streams_reader.c

```diff
@@ -957,99 +958,253 @@ static bool StreamReader_IsStream(RedisModuleKey *kp){
-        LockHandler_Release(staticCtx);
-
-        bool ret = StreamReader_VerifyCallReply(staticCtx, reply, "Failed scanning keys on background", "warning");
-        RedisModule_Assert(ret);
-
-        RedisModule_Assert(RedisModule_CallReplyType(reply) == REDISMODULE_REPLY_ARRAY);
+typedef struct ScanCtx {
+    Gears_list *keys;
+    char *pattern;
+} ScanCtx;
 
-        RedisModule_Assert(RedisModule_CallReplyLength(reply) == 2);
-
-        RedisModuleCallReply *cursorReply = RedisModule_CallReplyArrayElement(reply, 0);
-
-        RedisModule_Assert(RedisModule_CallReplyType(cursorReply) == REDISMODULE_REPLY_STRING);
-
-        RedisModuleString *cursorStr = RedisModule_CreateStringFromCallReply(cursorReply);
-        RedisModule_StringToLongLong(cursorStr, &cursor);
-        RedisModule_FreeString(NULL, cursorStr);
+static void StreamReader_ScanFreeKey(void* pd){
+    if (!pd) return;
+    RedisModuleString *keyname = pd;
+    RedisModule_FreeString(NULL, keyname);
+}
 
-        RedisModuleCallReply *keysReply = RedisModule_CallReplyArrayElement(reply, 1);
-        RedisModule_Assert(RedisModule_CallReplyType(keysReply) == REDISMODULE_REPLY_ARRAY);
-        if(RedisModule_CallReplyLength(keysReply) < 1){
-            RedisModule_FreeCallReply(reply);
-            continue;
+static int stringmatchlen(const char *pattern, int patternLen,const char *string, int stringLen, int nocase, int *skipLongerMatches, int nesting)
+{
+    /* Protection against abusive patterns. */
+    if (nesting > 1000) return 0;
+
+    while(patternLen && stringLen) {
+        switch(pattern[0]) {
+        case '*':
+            while (patternLen && pattern[1] == '*') {
+                pattern++;
+                patternLen--;
+            }
+            if (patternLen == 1)
+                return 1; /* match */
+            while(stringLen) {
+                if (stringmatchlen(pattern+1, patternLen-1,
+                            string, stringLen, nocase, skipLongerMatches, nesting+1))
+                    return 1; /* match */
+                if (*skipLongerMatches)
+                    return 0; /* no match */
+                string++;
+                stringLen--;
+            }
+            /* There was no match for the rest of the pattern starting
+             * from anywhere in the rest of the string. If there were
+             * any '*' earlier in the pattern, we can terminate the
+             * search early without trying to match them to longer
+             * substrings. This is because a longer match for the
+             * earlier part of the pattern would require the rest of the
+             * pattern to match starting later in the string, and we
+             * have just determined that there is no match for the rest
+             * of the pattern starting from anywhere in the current
+             * string. */
+            *skipLongerMatches = 1;
+            return 0; /* no match */
+            break;
```

Comment:
`break` after `return 0`?

Reply:
I guess it avoid warnings in some compilers, not sure. But as I said its a copy past from Redis.